h2o-extensions/target-encoder/src/test/java/ai/h2o/targetencoding/strategy/RGSWithFilteringTest.java

```diff
@@ -0,0 +1,269 @@
+import org.junit.BeforeClass;
+import org.junit.Test;
+import water.TestUtil;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import static org.junit.Assert.assertEquals;
+
+public class RGSWithFilteringTest extends TestUtil {
+
+  @BeforeClass
+  public static void setup() {
+    stall_till_cloudsize(1);
+  }
+
+  @Test
+  public void randomGridSearchWithFunctionsToFilterOutUnnecessaryGridItems() {
+    HashMap<String, Object[]> hpGrid = new HashMap<>();
+    hpGrid.put("_blending", new Boolean[]{true, false});
+    hpGrid.put("_noise_level", new Double[]{0.0, 0.01, 0.1});
+    hpGrid.put("_k", new Double[]{1.0, 2.0, 3.0});
+    hpGrid.put("_f", new Double[]{1.0, 2.0, 3.0});
+
+    TargetEncoderModel.TargetEncoderParameters parameters = new TargetEncoderModel.TargetEncoderParameters();
+
+    GridSearch.SimpleParametersBuilderFactory simpleParametersBuilderFactory = new GridSearch.SimpleParametersBuilderFactory();
+
+    HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria hyperSpaceSearchCriteria = new HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria();
+
+
+    KeepOnlyFirstMatchFilterFunction blendingFilterFunction = new KeepOnlyFirstMatchFilterFunction((gridItem) -> {
+      Object mainHP = gridItem.get("_blending");
+      return mainHP instanceof Boolean && !(Boolean) mainHP; // TODO handle cases when type is not that expected
+    });
+
+    Function<Map<String, Object>, Boolean> strictFilterFunction = gridItem -> {
+      return !((double) gridItem.get("_k") == 3.0 && (double) gridItem.get("_f") == 1.0);
+    };
+
+    ArrayList<Function<Map<String, Object>, Boolean>> filterFunctions = new ArrayList<>();
+    filterFunctions.add(strictFilterFunction);
+    filterFunctions.add(blendingFilterFunction);
+
+    HyperSpaceWalker.RandomDiscreteValueWalker<TargetEncoderModel.TargetEncoderParameters> walker =
+            new HyperSpaceWalker.RandomDiscreteValueWalker<TargetEncoderModel.TargetEncoderParameters>(parameters,
+                    hpGrid,
+                    simpleParametersBuilderFactory,
+                    hyperSpaceSearchCriteria,
+                    filterFunctions);
+
+    HyperSpaceWalker.HyperSpaceIterator<TargetEncoderModel.TargetEncoderParameters> iterator = walker.iterator();
+
+    ArrayList<TargetEncoderModel.TargetEncoderParameters> evaluatedGridItems = new ArrayList<>();
+    while (iterator.hasNext(null)) {
+      TargetEncoderModel.TargetEncoderParameters targetEncoderParameters = iterator.nextModelParameters(null);
+      if( targetEncoderParameters != null) { // we might have had next element but it can be filtered out by ffiltering unctions
+        evaluatedGridItems.add(targetEncoderParameters);
+      }
+    }
+    System.out.println("\nAll grid items after applying filtering functions:");
+    evaluatedGridItems.forEach(System.out::println);
+
+    assertEquals(54 - (27 - 1) + 3 - 6, evaluatedGridItems.size());
```

Comment:
can we know where those pretty numbers come from, please?

Reply:
@seb-h2o would you ask me for an explanation if I just put `25` there? :) I wanted to make it clearer actually where result is coming from. I thought it is obvious. Explaining it with word would take a lot of words though :) Will try to make it more clear

Reply:
no, it's just that when I read the test I don't know where they're coming from: simplest is to add a comment where those numbers are "created", like I suggested for the other test below.